miner/payload_building.go

```diff
@@ -209,23 +233,62 @@ func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
 		txs:         args.Transactions,
 		gasLimit:    args.GasLimit,
 	}
+
 	empty := w.getSealingBlock(emptyParams)
-	if empty.err != nil {
-		return nil, empty.err
-	}
+	return empty
+}
 
-	// Construct a payload object for return.
-	payload := newPayload(empty.block, args.Id())
+// buildPayload builds the payload according to the provided parameters.
+func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
+	// In OP-Stack, the "empty" block is constructed from provided txs only, i.e. no tx-pool usage.
 	if args.NoTxPool { // don't start the background payload updating job if there is no tx pool to pull from
+		// TODO: Do we also want to move the empty block generation in the NoTxPool case
+		//    into the goroutine below? In the current version and NoTxPool case,
+		//    we synchronously build the empty block, but when we do have a tx pool,
+		//    we build even the first block async. This behavior is less consistent,
+		//    but makes payload resolution easier.
+		//    On 2nd thought, I think it makes sense to keep this behavior because we use the NoTxPool case
+		//    when the block's timestamp is out of the sequencer drift window, so getPayload
+		//    can then immediately be called after forkchoiceUpdated.
+		empty := w.emptyBlock(args)
+		if empty.err != nil {
+			return nil, empty.err
+		}
+		payload := newPayload(empty.block, args.Id())
 		// make sure to make it appear as full, otherwise it will wait indefinitely for payload building to complete.
 		payload.full = empty.block
 		payload.fullFees = empty.fees
 		return payload, nil
 	}
 
+	fullParams := &generateParams{
```

Comment:
It's not ideal to pull the definition of `fullParams` out of the goroutine, because it increases the diff.

But we would miss parameter validation otherwise, so need to do this.

I caught this because the test `TestInvalidPayloadTimestamp` was failing.